cas-server-support-x509/src/main/java/org/jasig/cas/adaptors/x509/authentication/handler/support/LdapResourceCRLFetcher.java

```diff
@@ -0,0 +1,223 @@
+
+    /** The connection configuration to ldap.*/
+    protected final ConnectionConfig connectionConfig;
+
+    /** Search request to find the attribute in the ldap tree.*/
+    protected final SearchRequest searchRequest;
+
+    /** The ldap provider, defaults to {@link DefaultConnectionFactory#getProvider()}. */
+    protected Provider provider = DefaultConnectionFactory.getDefaultProvider();
+
+    /** The connection factory to prep for connections. **/
+    protected ConnectionFactory connectionFactory;
+
+    private AbstractConnectionPool connectionPool;
+
+    /**
+     * Instantiates a new Ldap resource cRL fetcher.
+     * initializes the default certificate revocation attribute
+     * to be {@link #DEFAULT_CERTIFICATE_REVOCATION_LIST_ATTRIBUTE}.
+     * @param searchRequest the search request
+     * @param connectionConfig the connection config
+     */
+    public LdapResourceCRLFetcher(@NotNull final SearchRequest searchRequest,
+                                  @NotNull final ConnectionConfig connectionConfig) {
+        this(searchRequest, connectionConfig,
+                DEFAULT_CERTIFICATE_REVOCATION_LIST_ATTRIBUTE);
+    }
+
+    /**
+     * Instantiates a new Ldap resource cRL fetcher.
+     *
+     * @param searchRequest the search request
+     * @param connectionConfig the connection config
+     * @param certificateRevocationListAttributeName the certificate revocation list attribute name
+     */
+    public LdapResourceCRLFetcher(@NotNull final SearchRequest searchRequest,
+                                  @NotNull final ConnectionConfig connectionConfig,
+                                  @NotNull final String certificateRevocationListAttributeName) {
+        this.certificateRevocationListAttributeName = certificateRevocationListAttributeName;
+        this.connectionConfig = connectionConfig;
+        this.searchRequest = searchRequest;
+    }
+
+    @Override
+    protected X509CRL fetchInternal(final Object r) throws Exception {
+        if (r.toString().toLowerCase().startsWith("ldap")) {
+            return fetchCRLFromLdap(r);
+        }
+        return super.fetchInternal(r);
+    }
+
+    public void setProvider(final Provider provider) {
+        this.provider = provider;
+    }
+
+    public void setConnectionPool(final AbstractConnectionPool connectionPool) {
+        this.connectionPool = connectionPool;
+    }
+
+    /**
+     * Downloads a CRL from given LDAP url
+     *
+     * @param r the resource that is the ldap url.
+     * @return the x 509 cRL
+     * @throws Exception if connection to ldap fails, or attribute to get the revocation list is unavailable
+     */
+    protected X509CRL fetchCRLFromLdap(final Object r) throws Exception {
+        Connection connection = null;
+        try {
+            final String ldapURL = r.toString();
+            logger.debug("Fetching CRL from ldap {}", ldapURL);
+
+            prepareConnectionFactory(ldapURL);
```

Comment:
Worked on supporting CAC [http://en.wikipedia.org/wiki/Common_Access_Card] a few months back. 
* You won't know what URL you are going to need to pull from until you read a card, and depending on the level you want to support, you will need the ability to pull from dozens of URLs. Your notification of a new host is when you read it from the card.
* Your PKI is configured with the highest CA cert you trust. This can be at the department level all the way to the full DoD. If you want to get fancy, you can also include other sections of the US government that use smart cards, which will add more points you'd have to support.
* chances are you are not going to need to deal with SSL since all the artifacts are signed. That's to say, you don't need to trust the connection, and in fact most resources are on unencrypted channels. You trust the artifacts at those locations because they are signed with a certificate you trust.
* you don't want to leave a connection open to a system serving CAC CRLs, especially as a non-US government or military entity

The best we found was to do something like [https://svn.apache.org/repos/asf/cxf/tags/cxf-2.4.4/distribution/src/main/release/samples/sts_issue_operation/src/main/java/demo/sts/provider/cert/CRLVerifier.java]. We had to work around the fact that systems come and go without notice, as well (e.g., one department stopped hosting the file on http).

I'm currently under NDA, so I can't go into much more detail about the implementation. I'll say, this was under CAS 4 and ldaptive added way too much noise to be tenable; it was much easier to just use `javax.naming`. It might be a different story if you are deploying for a single institution/organization, though.

Hope this helps.